### ISSUE TITLE
Issue with sourcemap sources paths with relative files and base

### ISSUE THREAD

**Issue by jayschab** [jonathanepollack/gulp-minify-css#89]
Hi there.

There seem to be many doubts and issues related to `gulp-sass` and sourcemaps in longer pipelines. So let me preface by saying I hope I'm not adding to the confusion :smile: 

My problem is getting sourcemaps to work with `gulp-sass` and this plugin. Here is my gulp file:

````javascript
var gulp         = require('gulp');  //gulp@3.8.11
var config       = require('../config').sass;
var merge = require('merge');
var replace = require('gulp-replace');
var gIf = require('gulp-if');

// Tools
var sass         = require('gulp-sass');  //gulp-sass@1.3.3
var sourcemaps   = require('gulp-sourcemaps'); //gulp-sourcemaps@1.5.2
var autoprefixer = require('gulp-autoprefixer'); //gulp-autoprefixer@2.1.0
var minifyCss = require('gulp-minify-css'); //gulp-minify-css@1.1.0

// List of browser string syntax for autoprefixer: https://github.com/ai/browserslist#queries
var supportedBrowsers = ['Chrome >= 11', 'Firefox >= 6', 'Explorer >= 10', 'Opera >= 5', 'Safari >= 5'];

gulp.task('sass', function () {
  var bowerIncludePaths = config.bowerIncludes.map(function(path) { return config.bowerDir + "/" + path; });
  var topLevelFiles = config.topLevelFiles.map(function(name) { return config.srcDirectory + name + ".scss"; });

  return gulp.src( topLevelFiles, {base: config.srcDirectory} )
    .pipe(sourcemaps.init())
    .pipe( sass( merge(config.settings, { includePaths: bowerIncludePaths })) )
    .pipe( minifyCss( ) ) // now minify
    .pipe(sourcemaps.write(".", {debug: true}))
    .pipe(gulp.dest(config.dest))
});
````

So `base` ends up being `./gulp/assets/stylesheets` and `topLevelFiles` paths like `./gulp/assets/stylesheets/application.scss'` and bowerIncludePaths paths like `./bower_components/bootstrap-sass/vendor/assets/stylesheets`

After running `gulp-sass` it stores sources relative to `base` so for example `/path/to/project/gulp/assets/stylesheets/vendor/example.scss` is stored under `vendor/example.scss` (the code for this is [here](https://github.com/dlmanning/gulp-sass/blob/v1.3.3/index.js#L53)).

When running `gulp-minify-css` I get an issue that the paths are made relative to the root again. On [line 43](https://github.com/jonathanepollack/gulp-minify-css/blob/v1.1.0/index.js#L43) we have the code:

````javascript
return path.relative(file.base, src === '$stdin' ? file.path : src);
````

Here `file.base` is still `./gulp/assets/stylesheets`, `src` is the relative path from earlier (`vendor/example.scss` for example). So the call becomes:

````javascript
return path.relative('./gulp/assets/stylesheets', 'vendor/example.scss');
=>  '../../../vendor/example.scss'
````

So when `gulp-sourcemaps` tries to load the sources for writing I get the errors (with `debug: true`):

````
gulp-sourcemap-write: No source content for '../../../vendor/example.scss'
gulp-sourcemap-write: source file not found: /path/to/project/vendor/example.scss
````

The only way I was able to get it to work was to change line 43 to be more similar to `gulp-sass` 
````javascript
   var absPath = src === '$stdin' ? file.path : path.resolve(file.base, src)
   return path.relative(file.base, absPath);
````

If it looks like this change makes sense I can make a PR. Does this looks like a reasonable solution? Or is this plugin doing the correct thing and the problem is in one of the other plugins?

**Comment by shinnn** [comment]
Hmm, I downloaded [your fork](https://github.com/jonathanepollack/gulp-minify-css/commit/2fe69937871a061f596a2cee2144ee75861370c5) and ran the test (`mocha`), but it failed.

**Comment by AlexanderOMara** [comment]
This looks like it's the same issue I'm having. It appears gulp-minify-css is changing the paths when it comes after another plugin in the sourcemap chain. Here is a test case using autoprefix, but the same thing happens with other plugins.

**gulpfile.js**
```js
'use strict';

var gulp = require('gulp');
var gulpMinifyCSS = require('gulp-minify-css');
var gulpSourcemaps = require('gulp-sourcemaps');
var gulpAutoprefixer = require('gulp-autoprefixer');

gulp.task('css', function() {
	return gulp.src('./src/css/**/*.css')
		.pipe(gulpSourcemaps.init())
		.pipe(gulpMinifyCSS({processImport: false}))
		.pipe(gulpSourcemaps.write('.', {}))
		.pipe(gulp.dest('./css/'));
});

gulp.task('css-autoprefix', function() {
	return gulp.src('./src/css/**/*.css')
		.pipe(gulpSourcemaps.init())
		.pipe(gulpAutoprefixer({browsers: ['> 0%'], cascade: false}))
		.pipe(gulpMinifyCSS({processImport: false}))
		.pipe(gulpSourcemaps.write('.', {}))
		.pipe(gulp.dest('./css-prefix/'));
});

gulp.task('default', ['css', 'css-autoprefix']);
```

**src/css/main.css**
```css
body {
	color: grey;
}
```

If we run `$ gulp`, we get the following `.map` files.

**css/main.css.map**
```json
{"version":3,"sources":["main.css"],"names":[],"mappings":"AAAA,KACC,MAAO","file":"main.css","sourcesContent":["body {\n\tcolor: grey;\n}\n"],"sourceRoot":"/source/"}
```

**css-prefix/main.css.map**
```json
{"version":3,"sources":["../../main.css"],"names":[],"mappings":"AAAA,KACC,MAAA","file":"main.css","sourcesContent":[null],"sourceRoot":"/source/"}
```

Above we can see the path was changed from `main.css` to `../../main.css` when run through gulp-minify-css after another plugin.

I've narrowed down the issue to [this line](https://github.com/jonathanepollack/gulp-minify-css/blob/9eb95b23723d86acbb6e371d97efc21d28bdb44c/index.js#L43):

```js
return path.relative(file.base, src === '$stdin' ? file.path : src);
```

It appears that `file.path` is used when it is the first plugin in the chain, but `src` is used afterwards. Because `src` is relative (`'main.css'` in this case), `path.relative` expands it relative to the gulp directory.

The fix seems obvious, always use `file.path`, but since this seems intentional, what if the significance of this extra logic?

**Comment by shinnn** [comment]
Thanks, @AlexanderOMara.

I think this solves the problem. What do you think?

```js
return (src === '$stdin' || src === file.relative) ?
       file.relative :
       path.relative(file.base, src);
```

**Comment by AlexanderOMara** [comment]
@shinnn That solves the path issue, but I just noticed there is another issue, which this almost solves.

If we look at `css-prefix/main.css.map` from above, this issue causes `sourcesContent` to be null
```
..."sourcesContent":[null]...
```

With this patch, the test case above is fixed, and `css-prefix/main.css.map` now looks like this.

```json
{"version":3,"sources":["main.css"],"names":[],"mappings":"AAAA,KACC,MAAA","file":"main.css","sourcesContent":["body {\n\tcolor: grey;\n}\n"],"sourceRoot":"/source/"}
```

Unfortunately, the same cannot be said for gulp-less.

**gulpfile.js**

```js
'use strict';

var gulp = require('gulp');
var gulpLess = require('gulp-less');
var gulpMinifyCSS = require('gulp-minify-css');
var gulpSourcemaps = require('gulp-sourcemaps');
var gulpAutoprefixer = require('gulp-autoprefixer');

gulp.task('css', function() {
	return gulp.src('./src/css/**/*.css')
		.pipe(gulpSourcemaps.init())
		.pipe(gulpMinifyCSS({processImport: false}))
		.pipe(gulpSourcemaps.write('.', {}))
		.pipe(gulp.dest('./css/'));
});

gulp.task('css-autoprefix', function() {
	return gulp.src('./src/css/**/*.css')
		.pipe(gulpSourcemaps.init())
		.pipe(gulpAutoprefixer({browsers: ['> 0%'], cascade: false}))
		.pipe(gulpMinifyCSS({processImport: false}))
		.pipe(gulpSourcemaps.write('.', {}))
		.pipe(gulp.dest('./css-prefix/'));
});

gulp.task('less', function() {
	return gulp.src('./src/less/**/*.less')
		.pipe(gulpSourcemaps.init())
		.pipe(gulpLess())
		.pipe(gulpMinifyCSS({processImport: false}))
		.pipe(gulpSourcemaps.write('.', {}))
		.pipe(gulp.dest('./lesscss/'));
});

gulp.task('default', ['css', 'css-autoprefix', 'less']);
```

**src/less/main.less**
```css
body {
	color: grey;
}
```

`$ gulp` produces:

**lesscss/main.css.map**
```json
{"version":3,"sources":["../../main.less"],"names":[],"mappings":"AAAA,KACC,MAAA","file":"main.css","sourcesContent":[null],"sourceRoot":"/source/"}
```

This still has the `..."sourcesContent":[null]...` problem. I believe this is because `file.relative` is `'main.css'`, rather than `'main.less'`.

The following alternative almost works.

```js
return path.relative(file.base, file.path);
```

It produces the following `lesscss/main.css.map` file.
```json
{"version":3,"sources":["main.less","main.css"],"names":[],"mappings":"AAAA,KCCC,MAAA","file":"main.css","sourcesContent":["body {\n\tcolor: grey;\n}\n",null],"sourceRoot":"/source/"}
```
Now there are two entries, `"main.less","main.css"`. The first one now has the source for it, the second is `null`.

**Comment by clempat** [comment]
I get the same issue for the path either with the solution of @shinnn when file.relative should be set ?

**Comment by ddprrt** [comment]
Wondering if this might be an issue with [CleanCSS](https://github.com/jakubpawlowicz/clean-css)

**Comment by clempat** [comment]
I will take a look deeper tomorrow. I am wondering if it is connected to the root defined which give the relative path back... If u had the path of the css as root like `fileOptions.root = options.relativeTo || path.resolve(path.dirname(file.path));` Only one test fails... Concerning stylus but as I never use stylus I dont know why it triggers a problem...

**Comment by shinnn** [comment]
Maybe I found a solution. https://github.com/jonathanepollack/gulp-minify-css/commit/f1758cfe1993c46acd69f68b2e67728ca7496729

@AlexanderOMara @clempat @ddprrt Could you try the current `master`?

```
$ npm install jonathanepollack/gulp-minify-css
```

**Comment by AlexanderOMara** [comment]
Looks fixed to me.

**Comment by shinnn** [comment]
Fixed in v1.1.1.

@jayschab @AlexanderOMara @clempat Thanks for your awesome feedback.

**Comment by clempat** [comment]
Hi sorry, European time ^^... But yeah thanks, the test I wrote for my case pass now ^^... Here was the test: https://github.com/jonathanepollack/gulp-minify-css/compare/master...clempat:master

Thank you...
